orchestra/todos/import_export.py

```diff
@@ -0,0 +1,122 @@
+import csv
+import json
+
+from django.conf import settings
+from django.db import transaction
+from django.utils import timezone
+from tempfile import NamedTemporaryFile
+
+from orchestra.core.errors import SpreadsheetImportError
+from orchestra.google_apps.permissions import write_with_link_permission
+from orchestra.google_apps.service import Service
+from orchestra.google_apps.convenience import get_google_spreadsheet_as_csv
+from orchestra.models import TodoListTemplateImportRecord
+
+
+REMOVE_IF_HEADER = 'Remove if'
+SKIP_IF_HEADER = 'Skip if'
+
+
+def _write_template_rows(writer, todo, depth):
+    writer.writerow(
+        [json.dumps(todo.get('remove_if', [])),
+         json.dumps(todo.get('skip_if', []))] +
+        ([''] * depth) +
+        [todo.get('description', '')])
+    # `reversed` iteration because the JSON-serialized order of
+    # children is the opposite of the top-to-bottom order in the
+    # spreadsheet.
+    for child in reversed(todo.get('items', [])):
+        _write_template_rows(writer, child, depth + 1)
+
+
+def _upload_csv_to_google(spreadsheet_name, file):
+    service = Service(settings.GOOGLE_P12_PATH,
+                      settings.GOOGLE_SERVICE_EMAIL)
+    sheet = service.insert_file(
+        spreadsheet_name,
+        '',
+        settings.ORCHESTRA_TODO_LIST_TEMPLATE_EXPORT_GDRIVE_FOLDER,
+        'text/csv',
+        'application/vnd.google-apps.spreadsheet',
+        file.name
+    )
+    service.add_permission(sheet['id'], write_with_link_permission)
+    return 'https://docs.google.com/spreadsheets/d/{}'.format(sheet['id'])
+
+
+def export_to_spreadsheet(todo_list_template):
+    """
+    Recursively descend down the to-do tree of a `todo_list_template`
+    and add a row per to-do to a CSV, uploading it to a Google Sheet
+    when done. The first two columns of the sheet capture json-encoded
+    remove_if and skip_if logic. After that, the text in the `i`'th
+    column represents a to-do's title that is `i` levels nested in the
+    to-do tree.
+    """
+    with NamedTemporaryFile(mode='w+', delete=False) as file:
+        writer = csv.writer(file)
+        writer.writerow([REMOVE_IF_HEADER, SKIP_IF_HEADER])
+        _write_template_rows(writer, todo_list_template.todos, 0)
```

Comment:
Is there always a single root for a todo_list_template's items? `_write_template_rows()` seems to expect a single `todo`, whereas this function passes `todo_list_template.todos`.

Reply:
Yup! There's always a single root. `todo_list_template.todos` is shaped like this:
```
{
...
items: [{..., items: [...]}, {..., items: [...]}, ...]
...
}
```